src/dbnode/storage/index/block.go

```diff
@@ -884,6 +919,189 @@ func (b *block) addQueryResults(
+			batch = b.appendFieldAndTermToBatch(batch, field, term, iterateTerms)
+			if len(batch) < batchSize {
+				continue
+			}
+
+			batch, size, err = b.addAggregateResults(cancellable, results, batch)
+			if err != nil {
+				iter.Close()
+				return false, err
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			iter.Close()
+			return false, err
+		}
+
+		if err := iter.Close(); err != nil {
+			return false, err
+		}
+	}
+
+	// NB: it is safe to Close() a fieldsAndTermsIterator twice. We structure
+	// the code to rely on this intentionally here, to ensure we're releasing
+	// any held resources by the iterator.
+	if err := iter.Close(); err != nil {
+		return false, err
+	}
+
+	// Add last batch to results if remaining.
+	if len(batch) > 0 {
+		batch, size, err = b.addAggregateResults(cancellable, results, batch)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	exhaustive := !opts.LimitExceeded(size)
+	return exhaustive, nil
+}
+
+func (b *block) appendFieldAndTermToBatch(
+	batch []AggregateResultsEntry,
+	field, term []byte,
+	includeTerms bool,
+) []AggregateResultsEntry {
+	// NB(prateek): we make a copy of the (field, term) entries returned
+	// by the iterator during traversal, because the []byte are only valid per entry during
+	// the traversal (i.e. calling Next() invalidates the []byte). We choose to do this
+	// instead of checking if the entry is required (duplicates may exist in the results map
+	// already), as it reduces contention on the map itself. Further, the ownership of these
+	// idents is transferred to the results map, which either hangs on to them (if they are new),
+	// or finalizes them if they are duplicates.
+	var (
+		entry            AggregateResultsEntry
+		lastField        []byte
+		lastFieldIsValid bool
+	)
+	// relying on the fact that iterator traversal is in order, we can avoid creating duplicate
+	// entries for the same fields, by checking the last batch entry to see if the bytes are
+	// the same.
+	if len(batch) > 0 {
+		lastFieldIsValid = true
+		lastField = batch[len(batch)-1].Field.Bytes()
+	}
+	if lastFieldIsValid && bytes.Equal(lastField, field) {
+		entry = batch[len(batch)-1] // avoid alloc cause we already have the field
+	} else {
+		entry.Field = b.pooledID(field) // allocate id because this is the first time we've seen it
+	}
+
+	if includeTerms {
+		// terms are always new (as far we know without checking the map for duplicates), so we allocate
+		entry.Terms = append(entry.Terms, b.pooledID(term))
+	}
+
+	batch = append(batch, entry)
+	return batch
+}
+
+func (b *block) pooledID(id []byte) ident.ID {
+	data := b.bytesPool.Get(len(id))
+	data.IncRef()
+	data.AppendAll(id)
+	data.DecRef()
+	return b.idPool.BinaryID(data)
+}
+
+func (b *block) addAggregateResults(
+	cancellable *resource.CancellableLifetime,
+	results AggregateResults,
+	batch []AggregateResultsEntry,
+) ([]AggregateResultsEntry, int, error) {
+	// Checkout the lifetime of the query before adding results
```

Comment:
nit: period at the end of all these comments

Reply:
sure

Reply:
done